ENH: Add subspace tracking to ORNMF (consistent with ORPCA)

tjof2 opened the pull request:
The ORPCA matrix factorization function supports a SGD method. Using this allows the subspace to be tracked if it changes over time, which is useful for e.g. foreground-background separation in videos.

This change copies the same subspace tracking code over to ONMF, which currently assumes that when `X = WH^T + E`, `W` is stationary for the whole dataset. If your `nsamples` dimension is time, this allows for tracking variations in `W`.

Also, the ONMF algorithm hasn't been accessible from `s.decomposition()` since it was introduced to HyperSpy - this was presumably an oversight.

tjof2 committed:
Initial commit of SGD

Enables subspace tracking for ONMF, consistent with the method for ORPCA

tjof2 committed:
Fixed bug in momentum sgd method

tjof2 committed:
Access ONMF from decomposition()

tjof2 committed:
Updated docs

tjof2 committed:
Fixed typo

tjof2 commented:
This is ready for review

tjof2 committed:
Fixed typo and rolled back incorrect change

tjof2 committed:
Missed parameters for onmf()

tjof2 committed:
Passes relevant test

francisco-dlp commented:
The code looks good, but it'll be good to:

* [ ] Add more information about the feature in the User Guide in the Big Data section.
* [ ] Add a reference to the relevant scientific article
* [ ] Make the ORPCA and ONMF keywords for this feature consistent

It'll be also good to hear from @to266—the author of the ONMF code—about this.

tjof2 commented:
Thanks @francisco-dlp  - I'll work my way through your list!

tjof2 committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into enh_ornmf_stochastic

tjof2 commented:
I'll work on the documentation.

tjof2 committed:
Rename 'ONMF' to 'ORNMF'

This is for consistency with ORPCA, where O=Online, R=Robust

tjof2 committed:
Formatting and clarified docs

tjof2 commented:
> * [x]  Add more information about the feature in the User Guide in the Big Data section.
> * [X]  Add a reference to the relevant scientific article
> * [x]  Make the ORPCA and ONMF keywords for this feature consistent

tjof2 committed:
Consistency between ORPCA and ORNMF

Updated documentation, made kwargs similar

tjof2 committed:
Fix tests

tjof2 committed:
Add deprecation warning

tjof2 commented:
This is now ready for review again @francisco-dlp

tjof2 committed:
Missed comma

tjof2 committed:
Revert "Missed comma"

This reverts commit 8c5e10f6b0f6edc0cf71094377cb1a441581ef26.

tjof2 committed:
Update rpca.py

tjof2 committed:
Update rpca.py

tjof2 committed:
pep8

tjof2 committed:
File rename

tjof2 commented:
Or rather, ready for review once I fix the tests...

tjof2 committed:
Fix test and rename file

tjof2 committed:
Update test_ornmf.py

Added tests

tjof2 committed:
Duplicate fixed

dnjohnstone commented on doc/user_guide/mva.rst:
```diff
@@ -297,6 +302,42 @@ of components to keep. Setting this to a small number is recommended to keep
+   >>> s.decomposition(algorithm='ORNMF',
+                       output_dimension=3)
+
+As with ORPCA described above, the MomentumSGD method  is useful for scenarios
```
the Momentum Stochastic Gradient Descent (Momentum SGD) method...

tjof2 replied:
good idea

tjof2 replied:
I've defined MomentumSGD higher up now.

dnjohnstone commented on doc/user_guide/mva.rst:
```diff
@@ -297,6 +302,42 @@ of components to keep. Setting this to a small number is recommended to keep
+
+The default and MomentumSGD solvers assumes an l2-norm minimization problem,
+which can still be sensitive to very heavily corrupted data. A more robust
+alternative is available, although it is typically much slower than the l2 approach
```
A more robust alternative is available, although it is typically much slower.

dnjohnstone reviewed (approved):
I think this was basically just waiting for the docs to be checked - I've suggested two very minor changes that could be made easily and then I think this could be merged.

tjof2 committed:
Update mva.rst

As per comments

tjof2 commented:
> I think this was basically just waiting for the docs to be checked - I've suggested two very minor changes that could be made easily and then I think this could be merged.

Thanks @dnjohnstone for your suggestions!

tjof2 committed:
Merge branch 'RELEASE_next_minor' into enh_ornmf_stochastic

tjof2 commented:
I don't believe there is anything more to review here?

tjof2 committed:
Merge branch 'RELEASE_next_minor' into enh_ornmf_stochastic

tjof2 committed:
Update docs on PCA/NMF

tjof2 committed:
Additional documentation improvements to mva.rst

tjof2 committed:
Fix merge conflict

dnjohnstone commented:
@francisco-dlp @ericpre - what is this waiting for? Just for conflicts to be resolved?

tjof2 commented:
> @francisco-dlp @ericpre - what is this waiting for? Just for conflicts to be resolved?

I fixed the conflicts @dnjohnstone - hadn't noticed those, thanks - tests still pass

dnjohnstone commented on hyperspy/learn/ornmf.py:
```diff
@@ -144,49 +156,84 @@ def __init__(self, rank, lambda1=1., kappa=1., store_r=False,
-                _logger.warning("Normalization with an iterator is not"
-                                " possible, option ignored.")
+                _logger.warning(
+                    "Normalization with an iterator is not" " possible, option ignored."
```
Too many "" --> Should be: "Normalization with an iterator is not possible, option ignored."

dnjohnstone commented on hyperspy/learn/rpca.py:
```diff
@@ -100,20 +102,16 @@ def svd(X):
     if lambda1 is None:
-        _logger.warning("Threshold 'lambda1' is set to "
-                        "default: 1 / sqrt(nsamples)")
+        _logger.warning("Threshold 'lambda1' is set to " "default: 1 / sqrt(nsamples)")
```
Again with the many "", there are a few examples in this file, and I won't mark them all....  In general you've made quite many 2-line things 1 line in this PR, with no change to content, was that intentional and is everything still under the max. line length?

dnjohnstone replied:
In fact there are quite a few changes e.g. ' to " and regarding adding or removing brackets that look a bit like some editor you've used has made some auto changes? Do we want all of them?

tjof2 replied:
resolved most of them

dnjohnstone commented on hyperspy/learn/rpca.py:
```diff
@@ -128,7 +126,7 @@ def svd(X):
         for i in range(power + 1):
             Y2 = np.dot(L.T, np.dot(L, Y2))
-        Q, tmp = scipy.linalg.qr(Y2, mode='economic')
+        Q, tmp = scipy.linalg.qr(Y2, mode="economic")
```
Were these changes from ' to " also intentional? I'd always preferred ' so curious as to why if so?

tjof2 replied:
This is thanks to [black](https://black.readthedocs.io/en/stable/index.html)

dnjohnstone reviewed (changes requested):
Thanks @tjof2 - just having another quick look and I get the impression your editor might have made some auto-style changes, some of which look a bit off - could you just double check that please?

tjof2 committed:
resolve linebreaks

tjof2 committed:
additional fix

tjof2 commented:
> Thanks @tjof2 - just having another quick look and I get the impression your editor might have made some auto-style changes, some of which look a bit off - could you just double check that please?

@dnjohnstone  I've undone most of the ones I think you were referring too. It's down to using black vs autopep8.

tjof2 committed:
clarifying comment

tjof2 committed:
Missed warning for ONMF in lazy.py

tjof2 committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into enh_ornmf_stochastic

tjof2 commented:
CI failures are unrelated - all are due to `KeyError: 'PhotometricInterpretation'`

tjof2 committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into enh_ornmf_stochastic

tjof2 commented:
@francisco-dlp / @ericpre please can we get this merged soon? I'd like to address #1159 but that would be simpler with this already in.

If not, please let me know if there's anything outstanding. Tests are failing due to the known tifffile issue.

dnjohnstone commented:
> @francisco-dlp / @ericpre please can we get this merged soon? I'd like to address #1159 but that would be simpler with this already in.
> 
> If not, please let me know if there's anything outstanding. Tests are failing due to the known tifffile issue.

Just to say - I'm now +1 on merging this.

dnjohnstone reviewed (approved):
Thanks for modifications.

ericpre commented on hyperspy/learn/mva.py:
```diff
@@ -154,8 +156,8 @@ def decomposition(self,
         (X, E) : (numpy array, numpy array)
-            If 'algorithm' == 'RPCA_GoDec' or 'ORPCA' and 'return_info' is True,
-            returns the low-rank (X) and sparse (E) matrices from robust PCA.
+            If 'algorithm' == 'RPCA_GoDec', 'ORPCA', 'ORNMF' and 'return_info' is True,
```
Wondering if it would be sensible to have `'RPCA'` instead of `'RPCA_GoDec'`. To be consistent with `ORPCA` and `ORNMF`, `GoDec` should be a method and `RPCA` an algorithm? At the moment, there is only one RPCA method but they could be more in the future!

This should be considered in the context of https://github.com/hyperspy/hyperspy/issues/1678#issuecomment-312683352 too... If we want to try to more the arguments of the `blind_source_separation` and the `decomposition` method, we can keep it as it is and leave it for a separated PR and for 2.0.

tjof2 replied:
Makes sense to me, but perhaps should be addressed separately to this PR?

ericpre replied:
Agree.

tjof2 replied:
Great - I'll keep it in mind for the near future

ericpre reviewed (commented):
This looks good to me, I have left one comment.

I would leave it to @francisco-dlp to check, since he already started this review and it knows this much better than me. He should have some time fairly soon, so he should not that long before it get merged.

tjof2 committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into enh_ornmf_stochastic

tjof2 committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into enh_ornmf_stochastic

tjof2 committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into enh_ornmf_stochastic

francisco-dlp commented:
This LGTM and I don't have any further suggestions. However, codecov is not happy. This is mainly, but not only, because multiple warnings are not hit by the tests. Would it be possible to increase the coverage of the tests?

tjof2 commented:
@francisco-dlp  yes I'll take a look at that.

tjof2 committed:
Update ORNMF tests

tjof2 committed:
Update RPCA test coverage

tjof2 committed:
Fix broken test coverage here

tjof2 committed:
Dimensions are incorrect

tjof2 committed:
Fix regex comparison

tjof2 commented:
@francisco-dlp I've added tests to catch the `ValueError`s and `VisibleDeprecationWarning`s that are raised in `rpca.py` and `ornmf.py`.

tjof2 committed:
Update test to include ORNMF

tjof2 committed:
add extra tests for default args

tjof2 committed:
Missed method off

tjof2 committed:
Add missing check for output_dimension

tjof2 committed:
Revert "Add missing check for output_dimension"

This reverts commit c6933c771f243dca9cb5e89e797b5fa357962842.

tjof2 committed:
add missing test for output_dimension

tjof2 committed:
Fix merge conflict

tjof2 commented:
OK @francisco-dlp I've fixed the code coverage as requested!

francisco-dlp commented:
Could you update this with the latest Rnm so that the tests pass? Otherwise I'll have to go one by one to check if the only issue is the ``tifffile`` bug...

tjof2 commented:
@francisco-dlp it already is up-to-date, no? I merged in the following already:
```
commit dbb9821d5a9d41dd6a78af646a73c499cc0ceaa6 (upstream/RELEASE_next_minor, origin/RELEASE_next_minor, origin/HEAD, RELEASE_next_minor)
Merge: 68edb0733 4e6ec698d
Author: Francisco de la Peña <frandelapena@gmail.com>
Date:   Wed Apr 1 17:59:19 2020 +0200

    Merge pull request #2311 from dnjohnstone/copyright_update
    
    update to copyright 2020
```

francisco-dlp commented:
Yes, sorry for the noise.

tjof2 commented:
No problem - I think it's just the tests that are queued up. It should be good I think

francisco-dlp commented:
It looks that some ORNMF tests are failing. There are other tests failing (my fault), but you can fix them by merging the current Rnm

tjof2 commented:
Will take a look this morning!

tjof2 committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into enh_ornmf_stochastic

tjof2 committed:
Update tolerance for tests based on matrix size

tjof2 commented:
> It looks that some ORNMF tests are failing. 

@francisco-dlp  this is resolved for me locally now with the latest commit, let's wait for CI to finish, then I believe we're good to go?

> There are other tests failing (my fault), but you can fix them by merging the current Rnm

This is done.

tjof2 commented:
> let's wait for CI to finish

```
hyperspy\tests\mva\test_bss.py ....................................       [ 55%]
hyperspy\tests\mva\test_decomposition.py ................................ [ 56%]
hyperspy\tests\mva\test_export.py ..                                      [ 56%]
hyperspy\tests\mva\test_ornmf.py ..........                               [ 57%]
hyperspy\tests\mva\test_rpca.py ..................                        [ 57%]
```

Success! Done! :)

tjof2 commented:
> let's wait for CI to finish

```
hyperspy\tests\mva\test_bss.py ....................................       [ 55%]
hyperspy\tests\mva\test_decomposition.py ................................ [ 56%]
hyperspy\tests\mva\test_export.py ..                                      [ 56%]
hyperspy\tests\mva\test_ornmf.py ..........                               [ 57%]
hyperspy\tests\mva\test_rpca.py ..................                        [ 57%]
```

Success! Done! :) I tidied up the ORNMF tests to make them more consistent (i.e. always measuring against the same tolerance condition rather than hard-coded numbers that were different for each test)

tjof2 committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into enh_ornmf_stochastic

tjof2 committed:
Fix bug that means error component is always positive -> consistent with ORNMF

tjof2 commented:
Spotted a bug that's been there since the start, fixed in `dac66eb` - the sparse error component in ORPCA could only be positive before now, due to the use of `np.maximum()`. I've replaced this with the correct soft-thresholding operator for l1-norm. 

Tests pass locally, so I'll wait for the CI.

```
hyperspy/tests/mva/test_bss.py ....................................             [ 55%]
hyperspy/tests/mva/test_decomposition.py ...................................... [ 56%]
hyperspy/tests/mva/test_export.py ..                                            [ 56%]
hyperspy/tests/mva/test_ornmf.py ..........                                     [ 57%]
hyperspy/tests/mva/test_rpca.py ..................                              [ 57%]
```
